.github/workflows/ci.yml

```diff
@@ -94,36 +100,33 @@ jobs:
 
   build_and_push:
     name: Build docker image
-    needs: [ test, lint ]
+    if: (github.ref == 'refs/heads/main') || (github.event_name == 'release')
+
+    needs: [test, lint]
     runs-on: ubuntu-latest
     steps:
 
       - name: Checkout the repo
-        uses: actions/checkout@v2
+        uses: actions/checkout@v4
 
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v3
 
-      - name: Build image
-        run: docker build -t $IMAGENAME:latest .
-
       - name: Install doctl
-        if: github.ref == 'refs/heads/main'
         uses: digitalocean/action-doctl@v2
         with:
           token: ${{ secrets.DIGITALOCEAN_ACCESS_TOKEN }}
 
       - name: Log in to DO Container Registry
-        if: github.ref == 'refs/heads/main'
         run: doctl registry login --expiry-seconds 600
 
-      - name: Tag image with run id
-        if: github.ref == 'refs/heads/main'
-        run: docker tag $IMAGENAME:latest $IMAGENAME:${{ github.run_id }}
-
-      - name: Push image to DO Container Registry
-        if: github.ref == 'refs/heads/main'
-        run: docker push $IMAGENAME:${{ github.run_id }} && docker push $IMAGENAME:latest
+      - name: Build, tag, and publish image
+        env:
+          IMAGETAG: ${{ github.ref == 'refs/heads/main' && 'latest' || github.ref_name }}
+        run: |
+          docker build -t $IMAGENAME:$IMAGETAG .
+          docker tag $IMAGENAME:$IMAGETAG $IMAGENAME:${{ github.run_id }}
+          docker push -a $IMAGENAME
```

Comment:
I liked being able to see the time spent on the build separately, could we keep it as a separate step?